populator-machinery/controller.go

```diff
@@ -108,11 +114,36 @@ type controller struct {
 	recorder             record.EventRecorder
 	referenceGrantLister referenceGrantv1beta1.ReferenceGrantLister
 	referenceGrantSynced cache.InformerSynced
+	// Boolean flag which determines whether or not to use cloud provider specific data populate functions
+	useProviderImpl bool
+	// Data population function, invoked when the useProviderImpl variable is set to true
+	populate func(context.Context, *PopulatorParams) error
+	// Data population completeness check function, return true when data transfer gets completed.
+	// Invoked when the useProviderImpl variable is set to true
+	populateComplete func(context.Context, *PopulatorParams) (bool, error)
+}
+
+type PopulatorParams struct {
+	KubeClient   kubernetes.Interface
+	StorageClass *storagev1.StorageClass
+	PvcPrime     *corev1.PersistentVolumeClaim
+	Pvc          *corev1.PersistentVolumeClaim
+	Unstructured *unstructured.Unstructured
 }
 
 func RunController(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
 	gk schema.GroupKind, gvr schema.GroupVersionResource, mountPath, devicePath string,
 	populatorArgs func(bool, *unstructured.Unstructured) ([]string, error),
+) {
+	RunControllerV2(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix, gk, gvr, mountPath, devicePath, populatorArgs,
+		true, nil, nil)
+}
+
+func RunControllerV2(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
```

Comment:
Instead of calling it "RunControllerV2", which doesn't really say much about the semantics of the function, I would name this something more descriptive like "RunControllerWithCustomMethod"

Reply:
Also for public methods, add some comments so that it can be auto-generated in go docs.

Reply:
What about "RunControllerWithProviderSpecificMethod"? To align with the boolean flag name "useProviderImpl"

Reply:
Or maybe just "RunControllerWithMethod"? to contrast with running with a pod.

Reply:
I add a VolumePopulatorConfig as controller input and rename  RunControllerV2 to RunControllerWithConfig